File: Java/arrays/ProductOfArrayExceptSelf.java
/**
 * Problem: Product of Array Except Self
 * LeetCode: https://leetcode.com/problems/product-of-array-except-self/
 *
 * Approach:
 *  - Use prefix and suffix products.
 *  - First pass: store the prefix products (product of all numbers before current index).
 *  - Second pass (reverse): multiply by suffix products (product of all numbers after current index).
 *  - This avoids using division and ensures O(n) time complexity.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1) (excluding output array)
 */

public class ProductOfArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        // Step 1: Compute prefix products
        result[0] = 1;
        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        // Step 2: Compute suffix products and multiply
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= suffix;
            suffix *= nums[i];
        }

        return result;
    }
}

File: Python/arrays/Product_of_Array_Except_Self.py
"""
🧩 LeetCode #238 - Product of Array Except Self
------------------------------------------------
Given an integer array nums, return an array answer such that answer[i]
is equal to the product of all the elements of nums except nums[i].
Solve without using division and in O(n) time complexity.
"""

from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [1] * n

        # Compute prefix products
        prefix = 1
        for i in range(n):
            answer[i] = prefix
            prefix *= nums[i]

        # Compute suffix products and multiply
        suffix = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= suffix
            suffix *= nums[i]

        return answer
